pkg/pushsync/pushsync.go

```
@@ -77,59 +78,139 @@ func (ps *PushSync) Close() error {
+
+	// Select the closest peer to forward the chunk
+	peer, err := ps.peerSuggester.ClosestPeer(chunk.Address())
+	if err != nil {
+		// If i am the closest peer then store the chunk and send receipt
+		if errors.Is(err, topology.ErrWantSelf) {
+
+			// Store the chunk in the local store
+			_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+			if err != nil {
+				return fmt.Errorf("chunk store: %w", err)
+			}
+			ps.metrics.TotalChunksStoredInDB.Inc()
+
+			// Send a receipt immediately once the storage of the chunk is successfull
+			err = ps.sendReceipt(w, chunk.Address())
+			if err != nil {
+				return fmt.Errorf("send receipt: %w", err)
+			}
 			return nil
 		}
 		return err
 	}
 
-	// create chunk and store it in the local store
-	addr := swarm.NewAddress(ch.Address)
-	chunk := swarm.NewChunk(addr, ch.Data)
-	_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
-	if err != nil {
+	//Dont forward this chunk to the same peer as the received peer
+	if p.Address.Equal(peer) {
 		return err
 	}
 
-	// push this to the closest node too
-	peer, err := ps.peerSuggester.ClosestPeer(addr)
+	// Forward chunk to closest peer
+	streamer, err := ps.streamer.NewStream(ctx, peer, nil, protocolName, protocolVersion, streamName)
 	if err != nil {
-		if errors.Is(err, topology.ErrWantSelf) {
-			// i'm the closest - nothing to do
-			return nil
-		}
+		return fmt.Errorf("new stream: %w", err)
+	}
+	defer streamer.Close()
+
+	wc, rc := protobuf.NewWriterAndReader(streamer)
+
+	if err := ps.sendChunkDelivery(wc, chunk); err != nil {
+		return fmt.Errorf("send chunk: %w", err)
+	}
+	receiptRTTTimer := time.Now()
+
+	receipt, err := ps.receiveReceipt(rc)
+	if err != nil {
+		return fmt.Errorf("receive receipt: %w", err)
+	}
+	ps.metrics.ReceiptRTT.Observe(time.Since(receiptRTTTimer).Seconds())
+
+	// Check if the receipt is valid
+	if !chunk.Address().Equal(swarm.NewAddress(receipt.Address)) {
+		ps.metrics.InvalidReceiptReceived.Inc()
 		return err
 	}
```

This code block is returning the error which is nil. It should return an appropriate error stating that the receipt address is not correct.

---

good catch